README.md
# Bloonix agent daemon

## About Bloonix

For general information about Bloonix, questions, ratings or other things please visit our base repository:

* https://github.com/bloonix/bloonix


ChangeLog
bloonix-agent (0.78) released at 2017-08-10

    * Feature Register.pm: if only a ipv6 address is found
      and no ipv4 address then the ipv6 address is used
      for the registration.

bloonix-agent (0.77) released at 2016-09-21

    * Fixed an issue with the host registration.

bloonix-agent (0.76) released at 2016-09-14

    * Improved initialization of the agent if no host is configured.

bloonix-agent (0.75) released at 2016-07-07

    * Kicked facter as dependency.

bloonix-agent (0.74) released at 2016-06-10

    * Fixed: host_id comparison in Validate.pm.

bloonix-agent (0.73) released at 2016-05-21

    * Fixed: host_id == 0 is now only checked if host_id is defined.

bloonix-agent (0.72) released at 2016-05-19

    * Host ID 0 is ignored now.

bloonix-agent (0.71) released at 2016-04-04

    * Fixed bloonix-init-host and the call of systemctl.

bloonix-agent (0.70) released at 2016-04-04

    * Check /bin/systemctl instead of /usr/lib/systemd to
      determine if systemd is used.

bloonix-agent (0.69) released at 2016-03-29

    * Fixed systemctl errors.

bloonix-agent (0.68) released at 2016-03-28

    * Fixed systemd/sysvinit/upstart installation routines.

bloonix-agent (0.67) released at 2016-03-20

    * Implemented global option "agent_active_when".

bloonix-agent (0.66) released at 2016-03-11

    * Implemented host option "when".

bloonix-agent (0.65) released at 2015-11-22

    * Improved logging for registration.

bloonix-agent (0.64) released at 2015-11-20

    * New feature: register host at the bloonix webgui.

bloonix-agent (0.63) released at 2015-09-18

    * Dispatcher.pm updated, need to restart agent and set the dependency
      to Bloonix::Core v0.27.

bloonix-agent (0.62) released at 2015-09-18

    * Just fixed the docs in main.conf and set the default agents to 4.

bloonix-agent (0.61) released at 2015-09-18

    * Fixed max_concurrent_hosts calculation.

bloonix-agent (0.60) released at 2015-09-14

    * Did some performance improvements and implemented the parameter
      max_concurrent_hosts.
    * Implemented agent option "set_tags".

bloonix-agent (0.59) released at 2015-08-28

    * Moved postinst parts into bloonix-init-agent.

bloonix-agent (0.58) released at 2015-08-18

    * Moved all sudo files to /etc/sudoers.d.

bloonix-agent (0.57) released at 2015-08-18

    * Fixed %preun section in spec file.
    * Moved the creation of user bloonix into the core package.

bloonix-agent (0.56) released at 2015-08-06

    * Updated mk_arrays to mk_array_accessors.
    * License information for Strawberry Perl added.

bloonix-agent (0.55) released at 2015-06-16

    * The agent passes now the host and service id via options
      --bloonix-host-id and --bloonix-service-id by default to
      all bloonix plugins.

bloonix-agent (0.54) released at 2015-06-16

    * Implemented support for Windows.
    * Expand sudoers settings for bloonix.

bloonix-agent (0.53) released at 2015-04-22

    * Fixed parsing of command options.

bloonix-agent (0.52) released at 2015-04-14

    * Satellite configuration parameters implemented.

bloonix-agent (0.51) released at 2015-04-07

    * Fixed bloonix-init-host and implemented Bloonix::IO::SIPC.

bloonix-agent (0.50) released at 2015-03-21

    * Implemented Bloonix::IO::SIPC as replacement for Bloonix:REST.

bloonix-agent (0.49) released at 2015-03-09

    * Improved parsing of nagios statistics.
    * Fixed uninizialied value host_id in get_services().
    * An empty host list is now allowed.
    * Fixed bloonix-init-host: the server response is
      now parsed correctly.
    * Added a default sudoers file for the bloonix agent.

bloonix-agent (0.48) released at 2015-01-27

    * Fixed permissions of bloonix-agent.log.

bloonix-agent (0.47) released at 2015-01-26

    * Fixed: Can't call method "notice" on an undefined value
      at /usr/share/perl5/Bloonix/Agent.pm line 81

bloonix-agent (0.46) released at 2015-01-26

    * Fixed creation of bloonix-agent.log.

bloonix-agent (0.45) released at 2014-12-23

    * Fixed reload function.

bloonix-agent (0.44) released at 2014-12-23

    * Improved data handling.

bloonix-agent (0.43) released at 2014-12-23

    * Fixed VERSION in Worker.pm.
    * Fixed nagios/simple plugin handling.

bloonix-agent (0.42) released at 2014-12-23

    * Added parameter poll_interval. The default is 60s.

bloonix-agent (0.41) released at 2014-12-17

    * Improved the communication stack between parent and children.
    * Improved the check distribution to children.
    * Replaced the keyword "nagios" with "simple".

bloonix-agent (0.40) released at 2014-12-11

    * Fixed the benchmark functions.

bloonix-agent (0.39) released at 2014-12-03

    * Fixed parsing nagios argument list.

bloonix-agent (0.38) released at 2014-12-03

    * Kicked debug messages.
    * Log command and arguments of nagios with log level info.

bloonix-agent (0.37) released at 2014-11-30

    * Changed the boot facility.
    * Fixed sudo mapping for services.

bloonix-agent (0.36) released at 2014-11-16

    * Added use_sudo as global configuration parameter for the agent.
    * Fixed owner of all directories and files within /etc/bloonix.

bloonix-agent (0.35) released at 2014-11-08

    * Decreased the poll interval from 20 to 15 seconds because the
      the lowest interval of check_frequency=high is 15 seconds.

bloonix-agent (0.34) released at 2014-11-05

    * Increased the default timeout for checks from 30 to 60 seconds.

bloonix-agent (0.33) released at 2014-11-05

    * Fixed the init script and removed S20 or any other
      prefix from basename.

bloonix-agent (0.32) released at 2014-11-03

    * Updated the license information.

bloonix-agent (0.31) released at 2014-10-25

    * On systems with systemd STDOUT and STDERR are redirected
      to syslog by default. On sysvinit system it's possible to
      redirect STDOUT and STDERR to syslog.

bloonix-agent (0.30) released at 2014-10-24


perl/lib/Bloonix/Agent/Benchmark.pm
package Bloonix::Agent::Benchmark;

use strict;
use warnings;
use base qw(Bloonix::Accessor);

__PACKAGE__->mk_accessors(qw/get_hosts get_services/);

sub new {
    my ($class, $num) = @_;
    my $self = bless { todo => $num }, $class;

    if ($num ne "noexec-only") {
        ($self->{num_hosts}, $self->{num_services})
            = split /:/, $num;

        $self->create_hosts;
        $self->create_services;
    }

    return $self;
}

sub noexec_only {
    my $self = shift;

    return $self->{todo} eq "noexec-only";
}

sub create_hosts {
    my ($self, $num) = @_;
    $num ||= $self->{num_hosts};
    my @hosts;

    foreach my $n (1..$num) {
        push @hosts, {
            host_id => $n,
            password => $n
        };
    }

    $self->get_hosts(\@hosts);
}

sub create_services {
    my ($self, $num) = @_;
    $num ||= $self->{num_services};
    my @services;

    foreach my $n (1..$num) {
        push @services, {
            service_id => $n,
            agent_id => "localhost",
            command => "check-loadavg",
            command_options => [
               { option => "warning", value => "avg1:20" },
               { option => "critical", value => "avg1:50" }
            ]
        };
    }

    $self->get_services({
        status => "ok",
        data => {
            services => \@services,
            interval => 15
        }
    });
}

sub get_check_result {
    my ($self, $command) = @_;
    my @stdout;

    my $ipc = bless {
        stderr => [],
        timeout => 0,
        unknown => "",
        stdout => \@stdout,
        exitcode => 0
    }, "Bloonix::IPC::Cmd";

    if ($command eq "check-cpustat") {
        push @stdout, '{"status":"OK","stats":{"system":"0.84","quest":"0.00","softirq":"0.01","total":"4.91","idle":"95.09","nice":"0.11","irq":"0.00","steal":"0.00","other":"0.00","user":"3.82","iowait":"0.13"},"message":"total=4.91%, iowait=0.13%, system=0.84%, user=3.82%"}';
    } elsif ($command eq "check-diskusage") {
        push @stdout, '{"status":"OK","stats":{"usage":20004106240,"iusage":"297898","iusageper":"17","itotal":"1815072","total":29233254400,"freeper":27,"ifree":"1517174","free":7744155648,"usageper":"73","ifreeper":83},"message":"/: - usageper=73%, iusageper=17%"}';
    } elsif ($command eq "check-loadavg") {
        push @stdout, '{"status":"OK","stats":{"avg5":"0.13","avg1":"0.09","avg15":"0.19"},"message":"avg1=0.09, avg5=0.13, avg15=0.19"}';
    } elsif ($command eq "check-memstat") {
        push @stdout, '{"status":"OK","stats":{"swaptotal":1999630336,"swapusedper":"18.66","swapcached":28168192,"buffers":226463744,"dirty":9494528,"inactive":828805120,"memfree":211472384,"memused":"2955882496","pagetables":41922560,"committed":5598396416,"memusedper":"71.04","writeback":0,"active":1313333248,"slab":282632192,"swapfree":1626533888,"memtotal":4160913408,"vmallocused":25653248,"swapused":"373096448","memrealfree":"1205030912","cached":767094784,"mapped":61657088},"message":"memusedper=71.04%, swapusedper=18.66%"}';
    } elsif ($command eq "check-netstat") {
        push @stdout, '{"status":"OK","stats":{"failed":"0.00","recv_udp_errs":"0.00","close_wait":0,"close":6,"unknown":0,"closing":0,"last_ack":0,"sent_udp_pcks":"6.50","established":"131","active":"3.00","sent_resets":"0.50","syn_send":0,"syn_recv":0,"fin_wait2":6,"recv_resets":"0.50","recv_udp_pcks":"6.50","fin_wait1":0,"passive":"2.75","time_wait":158},"message":"established=131, active=3.00, passive=2.75"}';
    } elsif ($command eq "check-netstat-port") {
        push @stdout, '{"status":"OK","stats":{"established":"33","fin_wait2":0,"syn_recv":0,"syn_send":0,"close_wait":0,"fin_wait1":0,"close":0,"unknown":0,"time_wait":0,"closing":0,"last_ack":0},"message":"port 5432: established=33"}';
    } elsif ($command eq "check-open-files") {
        push @stdout, '{"status":"OK","stats":{"fh_alloc":"5248","nr_inodes":"183778","nr_dentry":"239448","fh_free":"0","nr_unused":"205299","nr_free_inodes":"64439","fh_max":"405080"},"message":"fh_real_free(98.70%), fh_alloc(5248), fh_free(0), fh_max(405080)"}';
    } elsif ($command eq "check-pgswstat") {
        push @stdout, '{"status":"OK","stats":{"pgpgin":"0.00","pgmajfault":"0.00","pgpgout":"18.00","pswpin":"0.00","pgfault":"1553.00","pswpout":"0.00"},"message":"pgpgin=0.00, pgpgout=18.00, pswpin=0.00, pswpout=0.00"}';
    } elsif ($command eq "check-procstat") {
        push @stdout, '{"status":"OK","stats":{"count":"603","runqueue":"1","blocked":"0","running":"1","new":"42.50"},"message":"new=42.50, runqueue=1, count=603, blocked=0, running=1"}';
    } elsif ($command eq "check-sockstat") {
        push @stdout, '{"status":"OK","stats":{"sockets":"722","udp":"5","ipfrag":"0","tcp":"131","raw":"3"},"message":"sockets=722, tcp=131, udp=5, raw=3, ipfrag=0"}';
    } else {
        push @stdout, '{"status":"OK","message":"OK"}';
    }

    return $ipc;
}

1;
